File: dna.erl
-module(dna).
-export([count/1]).

count(Dna) ->
  count(Dna, {0,0,0,0}).

count([], {A,C,G,T}) ->
  io:fwrite("~B ~B ~B ~B~n", [A,C,G,T]);
count([$A|Dna], {A,C,G,T}) ->
  count(Dna, {A+1,C,G,T});
count([$C|Dna], {A,C,G,T}) ->
  count(Dna, {A,C+1,G,T});
count([$G|Dna], {A,C,G,T}) ->
  count(Dna, {A,C,G+1,T});
count([$T|Dna], {A,C,G,T}) ->
  count(Dna, {A,C,G,T+1}).

File: DNA.fs
#light

let rec count a c g t = function
  | 'A'::s -> count (a+1) c g t s
  | 'C'::s -> count a (c+1) g t s
  | 'G'::s -> count a c (g+1) t s
  | 'T'::s -> count a c g (t+1) s
  | _ -> printfn "%d %d %d %d" a c g t

let [|_;s|] = System.Environment.GetCommandLineArgs() 
count 0 0 0 0 (s.ToCharArray() |> Seq.toList)
